### PR TITLE
allow mutable references in const values when they point to no memory

### LINKED ISSUE #120450: Mutable reference protected by a mutex in static context is considered UB
I tried this code:

```rust
#![feature(const_mut_refs)]

use std::sync::Mutex;

static MUTABLE_REFERENCE_HOLDER: Mutex<&mut [u8]> = Mutex::new(&mut []);

fn main() {}
```

I expected to see this happen: This code should compile because the mutable reference is protected by a Mutex and therefore, it is safe to access.

Instead, this happened: I receive the following compiler error:
```
error[E0080]: it is undefined behavior to use this value
 --> src/main.rs:5:1
  |
5 | static MUTABLE_REFERENCE_HOLDER: Mutex<&mut [u8]> = Mutex::new(&mut []);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value at .data.value: encountered mutable reference in a `const` or `static`
  |
  = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
  = note: the raw bytes of the constant (size: 24, align: 8) {
              0x00 │ 00 00 00 00 00 __ __ __ ╾───────alloc3────────╼ │ .....░░░╾──────╼
              0x10 │ 00 00 00 00 00 00 00 00                         │ ........
          }
```

I am not entirely sure if I'm missing something here. Is this UB or not?
Since a mutable reference to a slice of u8 is Send the mutex is Send + Sync. Therefore, I assume this should be allowed? 

If this is in fact a mistake in the compiler, I'd be glad to try to provide a pull request with a fix. In my head the compiler need to look at the Sync trait and if it is implement by a type, this is not UB.

### Meta

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (6b4f1c5e7 2024-01-27)
binary: rustc
commit-hash: 6b4f1c5e782c72a047a23e922decd33e7d462345
commit-date: 2024-01-27
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/120450

The second commit is just some drive-by test suite cleanup.

r? @oli-obk

### CODE REVIEW

File: tests/ui/consts/const-mut-refs/const_mut_refs.rs [hunk 1]
```diff
@@ -28,6 +30,10 @@ const fn bazz(foo: &mut Foo) -> usize {
     foo.x
 }
 
+// Empty slices get promoted so this passes the static checks.
+// Make sure it also passes the dynamic checks.
+static MUTABLE_REFERENCE_HOLDER: Mutex<&mut [u8]> = Mutex::new(&mut []);
```
Comment (oli-obk) [inline tests/ui/consts/const-mut-refs/const_mut_refs.rs:35]: If I'm reading the code right, the following will now also be legal, right?

```rust
static mut FOO: &mut [u8] = &mut [42];
static BAR: Mutex<&mut [u8]> = Mutex::new(unsafe { &mut *FOO });
```

Comment (oli-obk) [reply tests/ui/consts/const-mut-refs/const_mut_refs.rs:35]: edit: ah no, we're still using a ZST check on top

Comment (oli-obk) [reply tests/ui/consts/const-mut-refs/const_mut_refs.rs:35]: I guess this is tested in mut_ref_in_final_dynamic_check

Comment (RalfJung) [reply tests/ui/consts/const-mut-refs/const_mut_refs.rs:35]: I'll add this test in that file, just to be extra safe.

Comment (RalfJung) [reply tests/ui/consts/const-mut-refs/const_mut_refs.rs:35]: Done that.

**Pull request merged**